xdsl/dialects/stablehlo.py

```diff
@@ -126,13 +128,69 @@ class SubtractOp(ElementwiseBinaryOperation):
     name = "stablehlo.subtract"
 
 
+@irdl_op_definition
+class TransposeOp(IRDLOperation):
+    """
+    Permutes the dimensions of `operand` tensor using `permutation` and produces a
+    `result` tensor. More formally, `result[result_index] = operand[operand_index]`
+    where `result_index[d] = operand_index[permutation[d]]`.
+
+    https://github.com/openxla/stablehlo/blob/main/docs/spec.md#transpose
+    """
+
+    name = "stablehlo.transpose"
+
+    ElementType = Annotated[Attribute, ConstraintVar("ElementType")]
+
+    operand = operand_def(TensorType[ElementType])
+    result = result_def(TensorType[ElementType])
+    permutation = attr_def(DenseArrayBase)
+
+    def __init__(
+        self, operand: SSAValue, permutation: DenseArrayBase, result_type: Attribute
```

Comment:
Would it be worth also supporting inputting a permutation as a sequence of ints?

Reply:
Others may disagree, but I think that it's not a big overhead for the caller to just do this bit themselves (`DenseArrayBase.create_dense_int_or_index(i64, (2, 1, 0))`)